src/binary_format.cc

```diff
@@ -101,67 +82,53 @@ void FactorGraph::load_variables(const std::string &filename) {
             << std::endl;
         std::abort();
     }
-    bool is_evidence = isevidence >= 1;
-    bool is_observation = isevidence == 2;
+    bool is_evidence = role_serialized >=
+                       1;  // TODO interpret as bit vector instead of number?
+    bool is_observation = role_serialized == 2;
     variable_value_t init_value = is_evidence ? initial_value : 0;
-
-    variables[id] = RawVariable(id, type_const, is_evidence, cardinality,
-                                init_value, init_value, is_observation);
-
+    variables[vid] = RawVariable(vid, dtype, is_evidence, cardinality,
+                                 init_value, init_value, is_observation);
     ++size.num_variables;
     if (is_evidence) {
       ++size.num_variables_evidence;
     } else {
       ++size.num_variables_query;
     }
   }
-  file.close();
 }
 
 void FactorGraph::load_factors(const std::string &filename) {
-  std::ifstream file;
-  file.open(filename.c_str(), std::ios::in | std::ios::binary);
-  variable_id_t variable_id;
-  weight_id_t weightid;
-  variable_value_t value_id;
-  int16_t type;
-  factor_arity_t arity;
-  variable_value_t should_equal_to;
+  std::ifstream file(filename, std::ios::in | std::ios::binary);
   while (file.good()) {
-    file.read((char *)&type, 2);
-    if (!file.read((char *)&arity, 4)) break;
-
-    type = be16toh(type);
-    arity = be32toh(arity);
-
+    uint16_t type;
+    factor_arity_t arity;
+    // read fields
+    read_be(file, type);
+    if (!read_be(file, arity)) break;
+    // register the factor
     factors[size.num_factors] =
         RawFactor(size.num_factors, -1, (factor_function_type_t)type, arity);
-
     for (factor_arity_t position = 0; position < arity; ++position) {
-      file.read((char *)&variable_id, 8);
-      file.read((char *)&should_equal_to, 4);
-      variable_id = be64toh(variable_id);
-      should_equal_to = be32toh(should_equal_to);
-
-      // check for wrong id
+      // read fields for each variable reference
+      variable_id_t variable_id;
+      variable_value_t should_equal_to;
+      read_be(file, variable_id);
+      read_be(file, should_equal_to);
       assert(variable_id < capacity.num_variables && variable_id >= 0);
-
-      // add variables to factors
+      // add to adjacency lists
       factors[size.num_factors].add_variable_in_factor(
           VariableInFactor(variable_id, position, should_equal_to));
       variables[variable_id].add_factor_id(size.num_factors);
     }
     size.num_edges += arity;
-
     switch (type) {
-      case (FUNC_AND_CATEGORICAL): {
-        uint32_t n_weights = 0;
-        file.read((char *)&n_weights, 4);
-        n_weights = be32toh(n_weights);
-
+      case FUNC_AND_CATEGORICAL: {
+        // weight references for categorical factors
+        factor_weight_key_t n_weights = 0;
```

Comment:
This is basically the number of weights *per factor*.  `num_weights_t` should be a safe upperbound for this, but I wanted to give a different name or more precise bound. `factor_weight_key_t` is basically the product of category values which determines how many weights a factor can have.